src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java

```diff
@@ -331,60 +279,294 @@ public ResourceKey<? extends Registry<T>> getRegistryKey() {
-        if (RegistryManager.ACTIVE.getRegistry(registryName) != null || this.registryFactory != null)
-            throw new IllegalStateException("Cannot create a registry for a type that already exists");
-
-        this.registryFactory = () -> sup.get().setName(registryName);
-        return new RegistryHolder<>(this.registryKey);
+        if (BuiltInRegistries.REGISTRY.containsKey(registryName) || this.customRegistry != null)
+            throw new IllegalStateException("Cannot create a registry that already exists - " + this.registryKey);
+        if (this.seenNewRegistryEvent)
+            throw new IllegalStateException("Cannot create a registry after NewRegistryEvent was fired");
+
+        RegistryBuilder<T> registryBuilder = new RegistryBuilder<>(this.registryKey);
+        consumer.accept(registryBuilder);
+        this.customRegistry = registryBuilder.create();
+        this.registryHolder = new RegistryHolder<>(this.registryKey);
+        this.registryHolder.registry = this.customRegistry;
+        return this.customRegistry;
     }
 
-    @SuppressWarnings("unchecked")
-    private void onFill(IForgeRegistry<?> registry) {
-        if (this.optionalTags == null)
+    private void addEntries(RegisterEvent event) {
+        if (!event.getRegistryKey().equals(this.registryKey)) {
             return;
+        }
+        this.seenRegisterEvent = true;
+        Registry<T> registry = event.getRegistry(this.registryKey);
+        this.aliases.forEach(registry::addAlias);
+        for (Entry<DeferredHolder<T, ? extends T>, Supplier<? extends T>> e : entries.entrySet()) {
+            event.register(this.registryKey, e.getKey().getId(), () -> e.getValue().get());
+            e.getKey().bind(false);
+        }
+    }
 
-        ITagManager<T> tagManager = (ITagManager<T>) registry.tags();
-        if (tagManager == null)
-            throw new IllegalStateException("The forge registry " + registry.getRegistryName() + " does not support tags, but optional tags were registered!");
-
-        Multimaps.asMap(this.optionalTags).forEach(tagManager::addOptionalTagDefaults);
+    private void addRegistry(NewRegistryEvent event) {
+        this.seenNewRegistryEvent = true;
+        if (this.customRegistry != null) {
+            event.register(this.customRegistry);
+        }
     }
 
-    private void addEntries(RegisterEvent event) {
-        if (event.getRegistryKey().equals(this.registryKey)) {
-            this.seenRegisterEvent = true;
-            for (Entry<RegistryObject<T>, Supplier<? extends T>> e : entries.entrySet()) {
-                event.register(this.registryKey, e.getKey().getId(), () -> e.getValue().get());
-                e.getKey().updateReference(event);
-            }
+    /**
+     * Specialized DeferredRegister for {@link Block Blocks} that uses the specialized {@link DeferredHolder}, {@link DeferredBlock} as the return type for {@link #register}.
```

Comment:
The reference to `DeferredHolder` is unnecessary and makes the javadoc less readable in my eyes.
```suggestion
     * Specialized DeferredRegister for {@link Block Blocks} that uses the specialized {@link DeferredBlock} as the return type for {@link #register}.
```

Reply:
Fixed in f2801cb37.